Updating Service AppProtocol to GA

robscott opened the pull request:
This is a PR for changes related to https://github.com/kubernetes/enhancements/issues/1507. I'm not sure what I need to add to update the feature gate related docs though.

/sig network

annajung commented:
/assign @kcmartin 
/milestone 1.20
/hold no k/k pr?

robscott commented:
Woops, sorry @annajung! I opened a k/k PR just after that: https://github.com/kubernetes/kubernetes/pull/96327

annajung commented:
Hey @robscott, sorry I missed your comment about needing help with docs related to feature gate. 

You need to make modification to `content/en/docs/reference/command-line-tools-reference/feature-gates.md` to add your stable feature under https://github.com/kubernetes/website/blob/master/content/en/docs/reference/command-line-tools-reference/feature-gates.md#feature-gates-for-graduated-or-deprecated-features. When adding your feature into the table, please also move the content from https://github.com/kubernetes/website/blob/master/content/en/docs/reference/command-line-tools-reference/feature-gates.md#feature-gates-for-alpha-or-beta-features into feature-gates-for-graduated-or-deprecated-features, maintaining the history for the alpha and beta in the graduated table. 

Thanks!

sftim commented on content/en/docs/concepts/services-networking/service.md:
```diff
@@ -203,8 +203,6 @@ described in detail in [EndpointSlices](/docs/concepts/services-networking/endpo
 
-{{< feature-state for_k8s_version="v1.19" state="beta" >}}
-
 The `AppProtocol` field provides a way to specify an application protocol for each Service port.
```
As this is graduating to GA, can we mention the type of the values this field takes? (eg: an arbitrary string? one from a set of valid values?) Are there common values people might use?

tengqm replied:
According to API spec,

```
This field follows standard Kubernetes label syntax. Un-prefixed names are reserved for IANA standard service names (as per RFC-6335 and http://www.iana.org/assignments/service-names). Non-standard protocols should use prefixed names such as 'mycompany.com/my-custom-protocol'. This is a beta field that is guarded by the `ServiceAppProtocol` feature gate and enabled by default.
```

According to the validation logic, this field must be a "qualified name".

sftim replied:
Ah, so we also need to move ServiceAppProtocol into the graduated section of https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/

The paragraph from the API spec I would copy in verbatim (minus the beta details). That way, it can also get localized.

tengqm replied:
Yes. The upstream PR is merged.

sftim replied:
@robscott would you be willing to expand the documentation? What's here now is rather light on detail.

tengqm replied:
Isn't this field name `appProtocol` instead?

robscott replied:
@tengqm Depends on if you're using Go or yaml :). I'll update it to use `appProtocol` since yaml is likely more common.

annajung commented:
/hold cancel 
k/k pr merged

irvifa reviewed (commented):
Do we need tech reviewer for this?
/lgtm

sftim reviewed (changes requested):
https://deploy-preview-24138--kubernetes-io-master-staging.netlify.app/docs/reference/command-line-tools-reference/feature-gates/ still lists ServiceAppProtocol as beta - that needs updating to record the graduation to stable.

/lgtm cancel

kcmartin commented:
Hi @robscott,

A friendly reminder that Docs "Ready for Review" Deadline is this Monday, Nov. 23rd for the 1.20 release.

Please reach out if you have any questions.

robscott commented:
@sftim @tengqm Thanks for the feedback! I've moved the feature gate down to the graduated section and updated the docs to include the detail on what values should be set for `appProtocol`.

robscott commented:
@kcmartin Thanks for the reminder! I believe this PR now qualifies as ready for review, let me know if I'm missed anything.

sftim commented on content/en/docs/concepts/services-networking/service.md:
```diff
@@ -203,10 +203,13 @@ described in detail in [EndpointSlices](/docs/concepts/services-networking/endpo
-{{< feature-state for_k8s_version="v1.19" state="beta" >}}
+The `appProtocol` field provides a way to specify an application protocol for
+each Service port. The value of this field is mirrored by corresponding
+Endpoints and EndpointSlice resources.
```
(nit)
```suggestion
Endpoint and EndpointSlice objects.
```

robscott replied:
Updated to use "objects", not as sure about "Endpoint". It looks like we've more consistently used "Endpoints" throughout docs from what I can tell. It's definitely confusing to have a plural resource name, but the alternative of "Endpoint" also feels confusing. Not sure what's best here but for now I've left it as "Endpoints".

sftim replied:
D'oh. The `kind` is _Endpoints_ plural isn't it. That was the right call.

robscott replied:
Haha yeah, it is super confusing, especially alongside the singular EndpointSlice name.

sftim commented on content/en/docs/concepts/services-networking/service.md:
```diff
@@ -203,10 +203,13 @@ described in detail in [EndpointSlices](/docs/concepts/services-networking/endpo
 
 ### Application protocol
 
-{{< feature-state for_k8s_version="v1.19" state="beta" >}}
```
You could leave this line in and mark it stable for v1.20 so people know to look at older docs if their cluster isn't bang up to date.

robscott replied:
Good call, added this back in, thanks!

sftim reviewed (commented):
/lgtm

irvifa commented:
Hi @robscott I think once you address the comments we’re good to go

sftim commented:
(I'd be happy to merge this as-is, then tweak later)

kcmartin commented on content/en/docs/concepts/services-networking/service.md:
```diff
@@ -203,10 +203,13 @@ described in detail in [EndpointSlices](/docs/concepts/services-networking/endpo
 
-{{< feature-state for_k8s_version="v1.19" state="beta" >}}
+The `appProtocol` field provides a way to specify an application protocol for
+each Service port. The value of this field is mirrored by corresponding
```
the corresponding

robscott replied:
Updated, thanks!

kcmartin reviewed (approved):
/lgtm

irvifa reviewed (commented):
/approve
